### PR TITLE
7702 latest spec update

### PR DESCRIPTION
**Pull request by sudeepdino008**
- implementation for delegated designation and other latest spec changes - https://eips.ethereum.org/EIPS/eip-7702
- the integration tests updated to latest execution-spec-tests release - pectra-devenet-3@v1.4.0

### CODE REVIEW

File: core/vm/instructions.go [hunk 1]
```diff
@@ -372,7 +372,7 @@ func opReturnDataCopy(pc *uint64, interpreter *EVMInterpreter, scope *ScopeConte
 
 func opExtCodeSize(pc *uint64, interpreter *EVMInterpreter, scope *ScopeContext) ([]byte, error) {
 	slot := scope.Stack.Peek()
-	slot.SetUint64(uint64(interpreter.evm.IntraBlockState().GetCodeSize(slot.Bytes20())))
+	slot.SetUint64(uint64(len(interpreter.evm.IntraBlockState().ResolveCode(slot.Bytes20()))))
```
Comment (yperbasis) [inline core/vm/instructions.go:375]: Please remove `IntraBlockState.GetCodeSize` because it's unused now.

File: core/state/intra_block_state.go [hunk 2]
```diff
@@ -450,9 +479,10 @@ func (sdb *IntraBlockState) Selfdestruct6780(addr libcommon.Address) {
 	if stateObject == nil {
 		return
 	}
-
 	if stateObject.newlyCreated {
-		sdb.Selfdestruct(addr)
+		if _, ok := types.ParseDelegation(sdb.GetCode(addr)); !ok {
```
Comment (yperbasis) [inline core/state/intra_block_state.go:483]: This interplay between EIP-7702 & EIP-6780 looks like a corner case that can be a source of tricky consensus bugs. Is it currently covered by the execution spec tests? We should also get it well specified in EIP-7702.

File: core/state_transition.go [hunk 3]
```diff
@@ -271,8 +271,12 @@ func (st *StateTransition) preCheck(gasBailout bool) error {
 			// libcommon.Hash{} means that the sender is not in the state.
 			// Historically there were transactions with 0 gas price and non-existing sender,
 			// so we have to allow that.
-			return fmt.Errorf("%w: address %v, codehash: %s", ErrSenderNoEOA,
-				st.msg.From().Hex(), codeHash)
+
+			// eip-7702 allows tx origination from accounts having delegated designation code.
+			if _, ok := st.state.GetDelegatedDesignation(st.msg.From()); !ok {
```
Comment (somnathb1) [inline core/state_transition.go:276]: Is this needed? I can see that `st.state.GetCodeHash()` does not return the   `codeHash` of the delegated code, even though `.GetCode()` returns the delegation string.

File: core/vm/eips.go [hunk 4]
```diff
@@ -320,3 +321,26 @@ func enable7516(jt *JumpTable) {
 		numPush:     1,
 	}
 }
+
+func enable7702(jt *JumpTable) {
+	jt[EXTCODECOPY].constantGas = params.WarmStorageReadCostEIP2929
```
Comment (somnathb1) [inline core/vm/eips.go:326]: Any reason to also update this constant gas which was also set in `enable2929`?

File: core/state/intra_block_state.go (hunk 2 above)
Comment (sudeepdino008) [reply core/state/intra_block_state.go:483]: it is covered in the tests (that's when I fixed it here).
I'll get a section added to the eip.

File: core/state_transition.go (hunk 3 above)
Comment (sudeepdino008) [reply core/state_transition.go:276]: GetCodeHash() does correspond to the delegation string; GetCode returns the delegation string.
Not sure if I understand the reasoning properly here.

A (EOA) delegates to B.
A makes a tx, its codeHash will be non-empty (and corresponds to delegation string). But since it's an EOA with delegated code, it should be allowed to make the tx.

File: core/vm/eips.go (hunk 4 above)
Comment (sudeepdino008) [reply core/vm/eips.go:326]: just making it explicit. No other reason.

File: core/vm/instructions.go (hunk 1 above)
Comment (sudeepdino008) [reply core/vm/instructions.go]: I instead added a ResolveCodeSize, since that would avoid loading code in memory (to find the length) when only the size but not the code is needed. Like in case of extcodesize.

File: core/state/intra_block_state.go (hunk 2 above)
Comment (sudeepdino008) [reply core/state/intra_block_state.go:494]: https://github.com/ethereum/EIPs/pull/8832 cc: @somnathb1

File: core/types/authorization.go [hunk 5]
```diff
@@ -20,39 +20,27 @@ import (
 type Authorization struct {
 	ChainID *uint256.Int      `json:"chainId"`
 	Address libcommon.Address `json:"address"`
-	Nonce   []uint64          `json:"nonce"`
-	V       uint256.Int       `json:"v"`
+	Nonce   uint64            `json:"nonce"`
+	YParity uint256.Int       `json:"v"`
```
Comment (yperbasis) [inline core/types/authorization.go:24]: The json should also be named "yParity" rather than "v".
v =  27 + yParity
or
v = 2*chainId + 35 + yParity
See [EIP-155](https://eips.ethereum.org/EIPS/eip-155)

File: core/vm/eips.go (hunk 4 above)
Comment (somnathb1) [reply core/vm/eips.go:326]: Better to just put down the changes only

File: core/state/intra_block_state.go [hunk 6]
```diff
@@ -282,6 +284,44 @@ func (sdb *IntraBlockState) GetCodeHash(addr libcommon.Address) libcommon.Hash {
...
+
+	return sdb.GetCode(addr)
+}
+
+func (sdb *IntraBlockState) ResolveCodeSize(addr libcommon.Address) int {
+	// eip-7702
+	size := sdb.GetCodeSize(addr)
+	if size == 23 {
+		// might be delegated designation
+		return len(sdb.ResolveCode(addr))
```
Comment (yperbasis) [inline core/state/intra_block_state.go:310]: Relying on `GetCodeSize` instead might be slightly more efficient

File: core/vm/instructions.go [hunk 7]
```diff
@@ -446,13 +447,18 @@ func opExtCodeCopy(pc *uint64, interpreter *EVMInterpreter, scope *ScopeContext)
 //	(6) Caller tries to get the code hash for an account which is marked as deleted,
 //
 // this account should be regarded as a non-existent account and zero should be returned.
+//
+//	(4) Caller tries to get the code hash of a delegated account, the result should be
```
Comment (yperbasis) [inline core/vm/instructions.go:451]: should be (7) instead of (4)

File: core/vm/evm.go [hunk 8]
```diff
@@ -194,7 +194,12 @@ func (evm *EVM) call(typ OpCode, caller ContractRef, addr libcommon.Address, inp
 	p, isPrecompile := evm.precompile(addr)
 	var code []byte
 	if !isPrecompile {
-		code = evm.intraBlockState.GetCode(addr)
+		code = evm.intraBlockState.ResolveCode(addr)
+		if depth == 0 && typ == CALL {
+			if dd, ok := evm.intraBlockState.GetDelegatedDesignation(addr); ok {
+				evm.intraBlockState.AddAddressToAccessList(dd)
```
Comment (yperbasis) [inline core/vm/evm.go:200]: Why do we need this? Isn't the delegated designation already added to the access list by `gasCallEIP7702`?

Comment (sudeepdino008) [reply core/vm/evm.go:200]: this scenario is when the `tx.to` has a DD. This is a top level "call" (depth==0) for which dynamic cost isn't calculated through gasCallEIP7702 (which is called only when call is done via evm opcode), and so DD is not added to access_list.
Also, the resolution here doesn't cost anything (the intrinsic cost is assumed to cover this). Only when resolution is while executing some message call evm opcode, does the warm/cold cost get added to dynamic cost.

Comment (yperbasis) [reply core/vm/evm.go:200]: I see. It's better to move this to `IntraBlockState.Prepare`, which handles the addition of the `tx.to` to the access list.

File: core/tracing/hooks.go [hunk 9]
```diff
@@ -293,6 +293,8 @@ const (
 	GasChangeCallStorageColdAccess GasChangeReason = 13
 	// GasChangeCallFailedExecution is the burning of the remaining gas when the execution failed without a revert.
 	GasChangeCallFailedExecution GasChangeReason = 14
+	// GasChangeDelegatedDesignation is the amount of gas that will be charged for resolution of delegated designation.
```
Comment (yperbasis) [inline core/tracing/hooks.go:296]: `core/tracing/gen_gas_change_reason_stringer.go` should be updated as well

File: core/state_transition.go [hunk 10]
```diff
@@ -348,60 +352,66 @@ func (st *StateTransition) TransitionDb(refunds bool, gasBailout bool) (*evmtype
...
+	verifiedAuthorities := make([]libcommon.Address, 0)
 	if len(auths) > 0 {
 		var b [33]byte
 		data := bytes.NewBuffer(nil)
 		for i, auth := range auths {
 			data.Reset()
 
-			// 1. authority recover
+			// 1. chainId check
+			if auth.ChainID != nil && auth.ChainID.Uint64() != 0 && auth.ChainID.Uint64() != st.evm.ChainRules().ChainID.Uint64() {
```
Comment (yperbasis) [inline core/state_transition.go:370]: ChainID is not limited to 64-bit, so this line should be
```
if auth.ChainID != nil && !auth.ChainID.IsZero() && auth.ChainID.CmpBig(st.evm.ChainRules().ChainID) != 0 {
```

**Review by yperbasis — CHANGES_REQUESTED** [review]

**Review by somnathb1 — APPROVED** [review]
`intra_block_state` can be optimized, other than that lgtm

**Comment by sudeepdino008** [comment]
> `intra_block_state` can be optimized, other than that lgtm

can you expand on this? I can consider it even if in a later PR.

**Comment by somnathb1** [comment]
> > `intra_block_state` can be optimized, other than that lgtm
> 
> can you expand on this? I can consider it even if in a later PR.

For instance, `ResolveCode` will call `GetCode` twice if there is no delegation. But that is  just nitpicking, you can leave it at this

**Review by yperbasis — APPROVED** [review]
